File: LeetCode-Java/Binary Tree Preorder Traversal.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> resultList = new ArrayList<Integer>();
        Stack<TreeNode> treeStack = new Stack<TreeNode>();
        while(root != null || !treeStack.empty())
        {
            if(root == null)
            {
                root = treeStack.pop();
            }
            else
            {
                resultList.add(root.val);
                treeStack.push(root.right);
                root = root.left;
            }
        }
        return resultList;
    }
}

File: Leetcode-cpp/BinaryTreePreorderTraversal.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
 #include <stack>
class Solution {
public:
    vector<int> preorderTraversal(TreeNode *root) {
        stack<TreeNode *> record;
        vector<int> result;
        record.push(root);
        while(!record.empty())
        {
            TreeNode *temp = record.top();
            record.pop();
            if(temp == NULL) continue;
            result.push_back(temp->val);
            record.push(temp->right);
            record.push(temp->left);
        }
        return result;
    }
};
